impl/src/main/java/org/ehcache/impl/internal/concurrent/ConcurrentHashMap.java

```diff
@@ -1163,6 +1164,40 @@ else if (t.removeTreeNode(p))
         return null;
     }
 
+  /**
+   * Remove and return all mappings for which the keys have the specified hashcode.
+   * @param keyHash the keys' hashcode.
+   * @return the removed mappings.
+   */
+  public final Map<K, V> removeAllWithHash(int keyHash) {
+      Map<K, V> invalidated = new HashMap<K, V>();
+
+      int hash = spread(keyHash);
+      for (Node<K, V>[] tab = table; ; ) {
+          Node<K, V> f;
+          int n, i;
+          if (tab == null || (n = tab.length) == 0 ||
+              (f = tabAt(tab, i = (n - 1) & hash)) == null)
+              break;
+          else if (f.hash == MOVED)
+              tab = helpTransfer(tab, f);
+          else {
+              Map<K, V> nodes = null;
+              synchronized (f) {
+                  if (tabAt(tab, i) == f) {
+                      nodes = nodesAt(f);
+                      setTabAt(tab, i, null);
+                      invalidated.putAll(nodes);
```

Comment:
Two map creations + putAll. I believe the code should pay more attention to that ... even if it means using the map in a side effect way as an argument.